azure_costs_exporter/allocated_vm_collector.py

```diff
@@ -0,0 +1,90 @@
+
+
+def _extract_resource_group(vm_id):
+    return vm_id.split('/')[4]
+
+class AzureAllocatedVMCollector(object):
+    """
+    Class to export information about allocated Azure virtual machines 
+    in Prometheus compatible format.
+    """
+
+    def __init__(self, application_id, application_secret, tenant_id, subscription_ids, metric_name):
+        """
+        Constructor. Access is granted to what Microsoft calls a service principal / Azure Active Directory
+        Application / app registration. Read more about this topic at
+        https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal.
+        This page will guide you how to obtain an application_id, and application_secret, and the tenant_id of
+        your Azure Active Directory. Please do not forget to grant "Reader" permissions to the app for all
+        subscriptions that you want to monitor.
+        
+        :param application_id: The application ID that is created during the Azure app registration.
+        :param application_secret: The application secret that is created during the Azure app registration.
+        :param tenant_id: The ID of your Azure Active Directory instance
+        :param subscription_ids: A _sequence_ of subscription IDs that shall be monitored. The application_id
+                                 required "Reader" permissions on each subscription.
+        :param metric_name: Name of the timeseries
+        """
+        self._metric_name = metric_name
+        self._subscription_ids = subscription_ids
+        self._credentials = ServicePrincipalCredentials(client_id=application_id,
+                                                        secret=application_secret,
+                                                        tenant=tenant_id)
+
+    def _create_gauge(self):
+        """
+        Create a gauge instance.
+
+        :return: prometheus_client gauge instance
+        """
+        description = "Number of virtual machines per Azure subscription, location, resource group, and vm size"
+        allocated_vms = GaugeMetricFamily(self._metric_name, description, labels=_BASE_COLUMNS)
+        return allocated_vms
+
+    def _collect_allocated_vms(self, allocated_vms):
+        rows = []
+
+        for subscription_id in self._subscription_ids:
+            compute_client = ComputeManagementClient(self._credentials, subscription_id)
+            for vm in compute_client.virtual_machines.list_all():
+                rows.append([subscription_id,
+                             vm.location,
+                             _extract_resource_group(vm.id),
+                             vm.hardware_profile.vm_size,
+                             1])
+
+        df = DataFrame(data=rows, columns=_ALL_COLUMNS)
+
+        groups = df.groupby(_BASE_COLUMNS).sum()
+
+        for name, value in groups.iterrows():
+            allocated_vms.add_metric(name, int(value.total))
```

Comment:
AFAIK, the first argument to `add_metric` is the list of label values. `name` seems to suggest something different. Am I miss-understanding something here?

Reply:
No, you are absolutely right. I lazily copy-pasted the code from the original exporter without thinking. Fixed everywhere.